### PR TITLE
Add File IO profiling

### PR DESCRIPTION
**Pull request by davidrogers-unity**

### CODE REVIEW

File: mono/profiler/mono-profiler-logging.c [hunk 1]
```diff
@@ -5633,7 +5644,10 @@ setup_user_options (const char *arguments) {
 	if (profiler->action_flags.track_stack) {
 		profiler->flags |= MONO_PROFILE_ENTER_LEAVE;
 	}
-	
+    if (profiler->action_flags.track_fileio) {
+        profiler->flags |= MONO_PROFILE_FILEIO;
+    }
```
Comment (joshpeterson) [inline mono/profiler/mono-profiler-logging.c]: Nit: Is there a tabs vs. spaces issue here? I've had problems in Mono code when my editor is setup up for spaces in Unity.

**Comment by joshpeterson** [comment]
Should we make the same changes in the unity-2017-02-staging branch for MonoBleedingEdge?

File: mono/profiler/mono-profiler-logging.c (hunk 1 above)
Comment (davidrogers-unity) [inline mono/profiler/mono-profiler-logging.c]: Possibly. I thought I'd gotten all of them.

**Comment by davidrogers-unity** [comment]
I was thinking of doing the bleeding edge one next. Is there an easy way to merge the changes to both? I haven't looked at that branch yet to see how well they transfer over.

**Comment by joshpeterson** [comment]
There is not an easy way - the two branches are rather different. It is probably easier to make the changes independently.

**Comment by davidrogers-unity** [comment]
Ok. unity-2017-02-staging is the one to go to then?

**Comment by joshpeterson** [comment]
Yes, that is the develop branch for MonoBleedingEdge.

File: mono/io-layer/io.c [hunk 2]
```diff
@@ -397,7 +398,10 @@ static gboolean file_read(gpointer handle, gpointer buffer,
 	if (bytesread != NULL) {
 		*bytesread = ret;
 	}
-		
+
+	if (mono_profiler_get_events () & MONO_PROFILE_FILEIO)
+		mono_profiler_fileio (NULL, NULL, 1, *bytesread);
```
Comment (joncham) [inline mono/io-layer/io.c]: It looks like `bytesread` can be NULL here. I think you want to pass `ret` instead.

File: mono/io-layer/io.c [hunk 3]
```diff
@@ -482,6 +486,10 @@ static gboolean file_write(gpointer handle, gconstpointer buffer,
 	if (byteswritten != NULL) {
 		*byteswritten = ret;
 	}
+
+	if (mono_profiler_get_events () & MONO_PROFILE_FILEIO)
+		mono_profiler_fileio (NULL, NULL, 0, *byteswritten);
```
Comment (joncham) [inline mono/io-layer/io.c]: Same here

File: mono/metadata/profiler-private.h [hunk 4]
```diff
@@ -36,6 +36,7 @@ void mono_profiler_code_transition (MonoMethod *method, int result) MONO_INTERNA
 void mono_profiler_allocation      (MonoObject *obj, MonoClass *klass) MONO_INTERNAL;
 void mono_profiler_monitor_event   (MonoObject *obj, MonoProfilerMonitorEvent event) MONO_INTERNAL;
 void mono_profiler_stat_hit        (guchar *ip, void *context) MONO_INTERNAL;
+void mono_profiler_fileio          (MonoObject *obj, MonoClass *klass, int kind, int count) MONO_INTERNAL;
```
Comment (joncham) [inline mono/metadata/profiler-private.h]: Why does this signature include a `MonoObject*` and a `MonoClass*`?

Comment (davidrogers-unity) [inline mono/metadata/profiler-private.h]: In theory to track them. I'm not sure if it's necessary though.

**Pull request merged**